과제1-1. let 제거해보기

hellomargo opened the pull request:
지금까지는 강의를 들으며, 리액트를 기초가 조금 밖에 없는 상태에서 시작하다보니 이론에서 부족함을 느꼈습니다.
강의문서내의 링크에 걸린 문서내용과 실습을 좀 더 하는 시간을 가졌습니다.
이제부터 과제를 진행하려고 합니다! 감사합니다.

hellomargo committed:
eslint 설치 및 설정
1강 개발환경설정, eslint설치 및 설정을 따라해보았다.

hellomargo committed:
babel 설치

hellomargo committed:
JSX(1-3강)강의들으며 index.js 작성
오늘까지 강의를 들으며 내용을 파악하는 시간을 가졌고, 이론을 학습했습니다.
이제 과제를 진행할 예정입니다.

hannut91 commented on src/index.js:
```diff
@@ -1,2 +1,54 @@
+  );
+}
+
+render();
```
<img width="193" alt="image" src="https://user-images.githubusercontent.com/14071105/166614683-b109bbaa-8b87-42c4-97eb-6837162b1c19.png">

코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.


이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님의 블로그 글을 참고해 주세요

See also

  - [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#파일-마지막에-자동으로-1개의-개행-문자-추가하기)

hannut91 reviewed (commented):
시작이 좋습니다 ㅎㅎ 과제의 의도를 파악해 보시고 과제를 진행해 주세요

hellomargo committed:
코드 마지막 줄에 빈 줄 추가

hellomargo committed:
과제1. 변수의 재할당 없이 문제 해결해보기

변수의 재할당을 하지 않기 위해서let이 아닌 const 사용

1. 변수 count값을 활용한 다른 변수 copyCount를 생성
2. const변수는 재할당이 안되므로, 배열의 pop, push를 사용해서 값을 교체.
3. handleClick함수내에서 render함수 실행시, 값을 넘겨줌

hellomargo committed:
eslint로 코드fix

hannut91 commented on src/index.js:
```diff
@@ -1,2 +1,74 @@
+  return element;
+}
+
+const count = 0;
+const copyCount = count;
+const countNum = [copyCount];
+
+function handleClick() {
+  if (countNum[0] === 0) {
+    countNum.pop();
+    countNum.push(1);
+  } else {
+    const resultCount = countNum[0] + 1;
+    countNum.pop();
+    countNum.push(resultCount);
+  }
+
+  render(countNum[0]);
+}
```
`const`는 값을 재할당할 수 없지만, Array가 참조하고 있는 값은 변경할 수 있다는 점을 이용해서 해결해 주셨네요 ㅎㅎ 마치 `Queue`를 사용하신 것 같기도 하고요 ㅎㅎ 

과제의 의도는 값을 변하지 않는(immutable)로 처리하는 것을 의도했던 것이었습니다. Array자체는 변경되지 않지만, 내부의 값이 변경되기 때문에, 예측하기 어려워집니다.

`render`의 값 `num`을 이용해서 화면을 그려주고, 클릭했을 때 현재 num의 값에 다가 무언가를 변경한 값을 다시 `render`로 전달하도록 한 번 해보세요

hellomargo committed:
과제1. 수정

- count -> countInit 변수로 변경하여 초기값으로 사용
- 클릭했을 때, render함수에 증가한 값을 전달

hannut91 commented on src/index.js:
```diff
@@ -1,2 +1,60 @@
+  return element;
+}
+
+const countInit = 0;
```
`countInit`이라는 이름은 카운트를 초기화해! 라고 얘기하는 것 같아요. `initialCount`는 어떨까요?

hellomargo committed:
과제1. 풀이강의 시청 후, 수정작업

- 상태를 오브젝트로 받도록 { } 코드 추가

hannut91 commented:
`test`폴더에 e2e테스트가 있는데, 요구사항에 맞게 수정해 보세요

hellomargo committed:
eslint, test폴더 e2e테스트 오류 수정

hellomargo commented:
> `test`폴더에 e2e테스트가 있는데, 요구사항에 맞게 수정해 보세요

수정해봤습니다!
npm run test:e2e 를 실행하여 표시되는 오류를 수정했는데, 함수가 정의되지 않았다고 해서 추가했습니다.
그런데, 다른 분들의 커밋내용을 보니 테스트 폴더는 수정하지 않은 것 같은데,
저는 왜 오류가 생기는 것인지 모르겠습니다... 설정에 누락된 것이 있을까요?

hannut91 commented:
음 CI에서는 통과된걸로 봐서 뭔가 설정문제 같네요.
